pkg/userop/middleware.go

```diff
@@ -82,102 +81,52 @@ func getGasPricesMiddleware(provider EthBackend, gasConfig GasConfig) middleware
 
 		logger.Debug("getting gas prices")
 
-		// Calculate maxPriorityFeePerGas
 		var maxPriorityFeePerGas *big.Int
-		if !op.MaxPriorityFeePerGas.IsZero() {
-			maxPriorityFeePerGas = op.MaxPriorityFeePerGas.BigInt()
-		} else {
-			var maxPriorityFeePerGasStr string
-			if err := provider.RPC().CallContext(ctx, &maxPriorityFeePerGasStr, "eth_maxPriorityFeePerGas"); err != nil {
-				return err
-			}
-
-			var ok bool
-			maxPriorityFeePerGas, ok = new(big.Int).SetString(maxPriorityFeePerGasStr, 0)
-			if !ok {
-				return fmt.Errorf("failed to parse maxPriorityFeePerGas: %s", maxPriorityFeePerGasStr)
-			}
-			logger.Debug("fetched maxPriorityFeePerGas", "maxPriorityFeePerGas", maxPriorityFeePerGas.String())
+		var maxFeePerGas *big.Int
 
-			// Increase maxPriorityFeePerGas to give user more
-			// flexibility in setting the gas price.
-			maxPriorityFeePerGas.Mul(
-				maxPriorityFeePerGas,
-				gasConfig.MaxPriorityFeePerGasMultiplier.BigInt())
+		chainId, err := provider.ChainID(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get chain ID: %w", err)
 		}
 
-		// Calculate maxFeePerGas
-		var maxFeePerGas *big.Int
-		if !op.MaxFeePerGas.IsZero() {
-			maxFeePerGas = op.MaxFeePerGas.BigInt()
-		} else {
-			// Get the latest block to read its base fee
-			block, err := provider.BlockByNumber(ctx, nil)
+		isPolygon := chainId.Cmp(big.NewInt(137)) == 0 || chainId.Cmp(big.NewInt(80002)) == 0
+
+		// for Polygon and Amoy, fetch from polygon gas station
+		if isPolygon {
+			maxFeePerGas, maxPriorityFeePerGas, err = getPolygonGasPrices(chainId)
 			if err != nil {
-				return err
+				logger.Error("failed to get gas prices from polygon gas station", "error", err)
 			}
-			blockBaseFee := block.BaseFee()
-			logger.Debug("fetched block base fee", "baseFee", blockBaseFee.String())
-
-			maxFeePerGas = new(big.Int).Mul(blockBaseFee, gasConfig.MaxFeePerGasMultiplier.BigInt())
-			maxFeePerGas.Add(maxFeePerGas, maxPriorityFeePerGas)
 		}
 
-		logger.Debug("calculated gas price", "maxFeePerGas", maxFeePerGas, "maxPriorityFeePerGas", maxPriorityFeePerGas)
-
-		op.MaxFeePerGas = decimal.NewFromBigInt(maxFeePerGas, 0)
-		op.MaxPriorityFeePerGas = decimal.NewFromBigInt(maxPriorityFeePerGas, 0)
+		// for other chains, or in case gas station is down, fetch from provider
+		if !isPolygon || err != nil {
```

Comment:
Looks like the `|| err != nil` part is redundant since all the previous errors have already been taken care of

Reply:
The idea is that if chain is Polygon and the call to gas station has NOT resulted into gas prices, we do NOT fail and get estimates from provider.

---

```go
if isPolygon {
	maxFeePerGas, maxPriorityFeePerGas, err = getPolygonGasPrices(chainId)
	if err != nil {
		logger.Error("failed to get gas prices from polygon gas station", "error", err)
	}
}
```

The error handling you are referring to is just to log that call to gas station was erroneous.

Reply:
Oh, now I see
Thanks for clarification